src/Args.hpp

```diff
@@ -0,0 +1,161 @@
+// Copyright (C) 2020 Joel Rosdahl and other contributors
+//
+// See doc/AUTHORS.adoc for a complete list of contributors.
+//
+// This program is free software; you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 3 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along with
+// this program; if not, write to the Free Software Foundation, Inc., 51
+// Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+
+#pragma once
+
+#include "system.hpp"
+
+#include "NonCopyable.hpp"
+#include "Util.hpp"
+
+#include "third_party/nonstd/optional.hpp"
+#include "third_party/nonstd/string_view.hpp"
+
+#include <deque>
+#include <string>
+
+class Args
+{
+public:
+  Args();
+  Args(const Args& other);
+  Args(Args&& other);
+
+  static Args from_argv(int argc, const char* const* argv);
+  static Args from_string(const std::string& command);
+  static nonstd::optional<Args> from_gcc_atfile(const std::string& filename);
+
+  Args& operator=(const Args& other);
+  Args& operator=(Args&& other);
+
+  bool operator==(const Args& other) const;
+  bool operator!=(const Args& other) const;
+
+  size_t size() const;
+  const std::string& operator[](size_t i) const;
+  std::string& operator[](size_t i);
+
+  // Accessor functions for the legacy API:
+  Args& operator*();
+  const Args* operator->() const;
+
+  // Return the argument list as a vector of raw string pointers. Callers can
+  // use `const_char<char* const*>(args.to_argv().data())` to get an array
```

Comment:
```suggestion
  // use `const_cast<char* const*>(args.to_argv().data())` to get an array
```

Reply:
> Did you by any chance test this on Windows?

I don't have (easy) access to Windows so I never test ccache there, except for the unit tests which I run in Wine from time to time, at least before a release. I ran them now and that actually exposed a unit test failure for this PR. Good suggestion!